Integrate inventory service

lesser1 opened the pull request:
Modify ProductView:
    - Display to buyer whether product variant is in stock or not.
    - Limit amount that can be ordered at once.
    - Display to employee and admin all the inventory status
      information of the product items of the currently displayed
      product variant.
    - Add toolbar button for opening the new RestockDialog.

Add RestockDialog:
    - With this dialog the user can add batches of product items to
      the inventory.

[The Gropius Issue](https://frontend.gropius.duckdns.org/components/9d12d6ac-e85a-407c-bdff-a8321fb75d3a/issues/47c93561-8c77-440b-b419-34843ce57978)

### Definition of Done
- [x] Requirements of the issue are met
- [x] Code has been reviewed
- [x] Code is documented
- [x] GraphQL related artefacts are documented

lesser1 committed:
Integrate inventory service:

Modify ProductView:
    - Display to buyer whether product variant is in stock or not.
    - Limit amount that can be ordered at once.
    - Display to employee and admin all the inventory status
      information of the product items of the currently displayed
      product variant.
    - Add toolbar button for opening the new RestockDialog.

Add RestockDialog:
    - With this dialog the user can add batches of product items to
      the inventory.

lesser1 committed:
Run Formatter

nk-coding commented on src/components/RestockDialog.vue:
```diff
@@ -0,0 +1,320 @@
+    const length =
+        props.maximumBatchSize !== undefined
+            ? props.maximumBatchSize
+            : defaultMaximumLength
```
why not use the default field when defining the prop?

lesser1 replied:
"Fixed" with [this commit](https://github.com/MiSArch/frontend/pull/37/commits/8cd627d1c7be8ddc783c9b60d64100b59ae5b640)

nk-coding commented on src/components/RestockDialog.vue:
```diff
@@ -0,0 +1,320 @@
+function resetViewModel(): boolean {
+    if (hasChanges.value) {
+        selectedProductVariantId.value = props.preselectedProductVariantId
+
```
I know newlines allow to separete stuff, but in this function, it feels a bit excessive, at least to me, is there some reason why everything has an empty line afterwards?

lesser1 replied:
"Fixed" with [this commit](https://github.com/MiSArch/frontend/pull/37/commits/ef62ca911e902237fdbed3424e2917a86b3b7ec6)

nk-coding commented on src/components/RestockDialog.vue:
```diff
@@ -0,0 +1,320 @@
+ * Resets the view model to its initial state.
+ * @returns Whether or not the view model was reset.
+ */
+function resetViewModel(): boolean {
```
do you need the return value anywhere?

lesser1 replied:
As of now, no... well, I'll remove it to keep things simple. [This commit](https://github.com/MiSArch/frontend/pull/37/commits/1862dd4ae43a39b18b082070af15974e20036c40) does it.

nk-coding commented on src/components/RestockDialog.vue:
```diff
@@ -0,0 +1,320 @@
+ * @returns Whether or not the view model was reset.
+ */
+function resetViewModel(): boolean {
+    if (hasChanges.value) {
```
is there a reason for tracking the changes?
to my understanding, this just tracks if resetting is necessary, however, resetting if there are no changes should do no harm and would probably simplify the code as change tracking would no longer be required

lesser1 replied:
Yes, I have implemented tracking routinely. Now that I think about it, I like the suggestion of simplifying the code by taking that away.

lesser1 replied:
[The commit that "fixes" it...](https://github.com/MiSArch/frontend/pull/37/commits/433d32ec2bff6aef7e49908e2ccc6cced6d0ccda)

nk-coding commented on src/components/RestockDialog.vue:
```diff
@@ -0,0 +1,320 @@
+ */
+async function restock() {
+    if (activeUserRoleIsEitherAdminOrEmployee.value && canRestock) {
+        const q = await pushErrorNotificationIfNecessary(() => {
```
what does `q` mean? was this a temporary variable name?

lesser1 replied:
Yes, forgot to clean it up. [There you go](https://github.com/MiSArch/frontend/pull/37/commits/a3ae4b7e7badf322bdec788162fa1feec2682eaa)

nk-coding commented on src/views/ProductView.vue:
```diff
@@ -578,4 +622,247 @@ function goToWishlists() {
+ * Computed property indicating whether there are product items in stock.
+ * @returns True if there are product items in stock, false otherwise.
+ */
+const inStock = computed(() => {
```
I know this is currently the only way how you can do this.
However this is also really inefficient, in particular with lots of product items.
Imho the `productItems` field should provide a `filter` option, so that you can just use `totalCount` to fetch this information.
Maybe open an issue on `inventory`?

lesser1 replied:
I agree! I will open an issue.

lesser1 replied:
[The issue](https://frontend.gropius.duckdns.org/components/ba383e91-738a-488f-baf4-4b5d8876bd74/issues/41d14a98-238e-42ad-b72c-14b6aa313beb)

nk-coding commented on src/views/ProductView.vue:
```diff
@@ -578,4 +622,247 @@ function goToWishlists() {
+ * it returns the default maximum value of 10.
+ */
+const maximumNumberOfProductItemsABuyerCanOrder = computed(() => {
+    const defaultValue = 10
```
constants somewhere in the code are generally not liked by Becker and co
maybe move to the top of the file as a global constant, or even introduce some constants / configuration file

lesser1 replied:
I moved it to the top of the file for now: b0e0e45dbda8c3316007e96ee48b807848fd4810

nk-coding commented on src/views/ProductView.vue:
```diff
@@ -578,4 +622,247 @@ function goToWishlists() {
+        numberOfProductItemsInStock.value !== undefined &&
+        numberOfReservedProductItems.value !== undefined
+    ) {
+        const numberOfAvailableProductItems =
```
a product item has only one state
so there cannot be any overlap between in storage and reserved ones
thus I think the correct calculation would be just taking the  `numberOfProductItemsInStock` ones

lesser1 replied:
-> 821d4dd7f52015e8a6f78424b916ca72ccf895a2

nk-coding reviewed (commented):
Did not test it yet, will do tomorrow

lesser1 committed:
Add default value to prop maximumBatchSize

lesser1 committed:
Remove unnecessary newlines in function

lesser1 committed:
Remove unnecessary return type / value of function

lesser1 committed:
Remove return statements and change tracking

lesser1 committed:
Remove temp variable

lesser1 committed:
Move const to top of file

This is a temporary fix, in the near future, I want to introduce
some sort of configurations file.

lesser1 committed:
Correct calculation of number of orderable items

lesser1 committed:
Remove if condition

that caused the following issue: The inventory status information
would only be requested if the user was an employee or an admin.
